pkg/segment/results/mresults/seriesresult.go

```diff
@@ -361,41 +361,28 @@ func ApplyRangeFunction(ts map[uint32]float64, function structs.Function) (map[u
 		delete(ts, timestamps[0])
 		return ts, nil
 	case segutils.Rate:
-		// Calculate the average rate (per-second rate) for each timestamp. E.g: to determine the rate for the current point with its timestamp,
-		// find the earliest point within the time window: [timestamp - time window, timestamp]
-		// Then, calculate the rate between that point and the current point
-		if len(sortedTimeSeries) == 0 {
-			return nil, nil
-		}
-
+		return evaluateRate(sortedTimeSeries, ts, timeWindow), nil
+	case segutils.IRate:
+		// Calculate the instant rate (per-second rate) for each timestamp, based on the last two data points within the timewindow
+		// If the previous point is outside the time window, we still need to use it to calculate the current point's rate, unless its value is greater than the value of the current point
 		var dx, dt float64
-		resetIndex := -1
 		for i := 1; i < len(sortedTimeSeries); i++ {
-			timeWindowStartTime := sortedTimeSeries[i].downsampledTime - timeWindow
-			preIndex := sort.Search(len(sortedTimeSeries), func(j int) bool {
-				return sortedTimeSeries[j].downsampledTime >= timeWindowStartTime
-			})
-
-			if i <= preIndex { // Can not find the second point within the time window
+			timeDff := sortedTimeSeries[i].downsampledTime - sortedTimeSeries[i-1].downsampledTime
```

Comment:
Are you supposed to use the downsampled time here for `irate` that you're doing? If so, you'll need to handle the case `timeDff == 0` so you don't divide by zero on line 388 when you do `ts[sortedTimeSeries[i].downsampledTime] = dx / dt`

Reply:
I think if there are elements with different indexes in sortedTimeSeries, timeDiff would not be the same because ts is a map and sortedTimeSeries was converted from that map.

https://github.com/siglens/siglens/blob/d42e43040c7e596fffb426e05455f49f6b78e337/pkg/segment/results/mresults/seriesresult.go#L289-L303